lib/Targets/AIETargetCDO.cpp

```diff
@@ -0,0 +1,1084 @@
+               << tileDMAInstRefStr(col, row, bdNum) << ", "
+               << tileLocStr(col, row) << ");\n";
+        output << "XAie_DmaSetLock(" << tileDMAInstRefStr(col, row, bdNum)
+               << ", "
+               << "XAie_LockInit(" << acqLockId << "," << acqValue << "),"
+               << "XAie_LockInit(" << relLockId << "," << relValue << "));\n";
+
+        if (0 == ndims) {
+          output << "XAie_DmaSetAddrLen(" << tileDMAInstRefStr(col, row, bdNum)
+                 << ", /* addrA */ "
+                 << "0x" << llvm::utohexstr(BaseAddrA + offsetA) << ", "
+                 << " /* len */ " << lenA << " * " << bytesA << ");\n";
+        } else {
+          generateXAieDmaSetMultiDimAddr(output, ndims, dims, col, row, bdNum,
+                                         BaseAddrA, offsetA, lenA, bytesA, "");
+        }
+
+        if (block.getNumSuccessors() > 0) {
+          Block *nextBlock = block.getSuccessors()[0]; // should have only one
+                                                       // successor block
+          int nextBdNum = blockMap[nextBlock];
+          output << "XAie_DmaSetNextBd(" << tileDMAInstRefStr(col, row, bdNum)
+                 << ", "
+                 << " /* nextbd */ " << nextBdNum << ", "
+                 << " /* enableNextBd */ 1);\n"; // TODO Check if br ^end: to
+                                                 // disable this?
+        }
+        if (foundBdPacket) {
+          output << "XAie_DmaSetPkt(" << tileDMAInstRefStr(col, row, bdNum)
+                 << ", " << packetStr(packetID, packetType) << ");\n";
+        }
+        output << "XAie_DmaEnableBd(" << tileDMAInstRefStr(col, row, bdNum)
+               << ");\n";
+        output << "XAie_DmaWriteBd(" << deviceInstRef << ", "
+               << tileDMAInstRefStr(col, row, bdNum) << ", "
+               << tileLocStr(col, row) << ", "
+               << " /* bd */ " << bdNum << ");\n";
+      }
+    }
+
+    for (auto &block : memOp.getBody()) {
+      for (auto op : block.getOps<DMAStartOp>()) {
+        int bdNum = blockMap[op.getDest()];
+
+        llvm::StringRef dmaDir = stringifyDMAChannelDir(op.getChannelDir());
+        int chNum = op.getChannelIndex();
+
+        output << "XAie_DmaChannelPushBdToQueue(" << deviceInstRef << ", "
+               << tileLocStr(col, row) << ", "
+               << "/* ChNum */" << chNum
+               << ", "
+               // TODO hack until physical dialect changes
+               << "/* dmaDir */ DMA_" << dmaDir << ", "
+               << "/* BdNum */" << bdNum << ");\n";
+        output << "XAie_DmaChannelEnable(" << deviceInstRef << ", "
+               << tileLocStr(col, row) << ", "
+               << "/* ChNum */ " << chNum
+               << ", "
+               // TODO hack until physical dialect changes
+               << "/* dmaDir */ DMA_" << dmaDir << ");\n";
+      }
+    }
+  }
+
+  // for (auto op : module.getOps<ExternalBufferOp>()) {
```

Comment:
`#if 0`?
Well, at least I cannot complains about this is purely uncommented opaque open-source software. ;-)